src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2133,11 +2131,13 @@ internal override bool KnownCircularStruct
                     else
                     {
                         var diagnostics = BindingDiagnosticBag.GetInstance();
+                        Debug.Assert(diagnostics.DiagnosticBag is not null);
                         var value = (int)CheckStructCircularity(diagnostics).ToThreeState();
 
                         if (Interlocked.CompareExchange(ref _lazyKnownCircularStruct, value, (int)ThreeState.Unknown) == (int)ThreeState.Unknown)
                         {
-                            AddDeclarationDiagnostics(diagnostics);
+                            DeclaringCompilation.AddUsedAssemblies(diagnostics.DependenciesBag);
```

Comment:
>AddUsedAssemblies [](http://example.com/codeflow?start=49&length=17)

I am not comfortable with pushing diagnostics and dependencies to the compilation at different times and through different caches. Instead, I suggest we switch ```CircularStructDiagnostics``` to a full ```BindingDiagnosticBag``` with dependencies bag backed by a concurrent set.